### ISSUE TITLE
[compiler-rt][tysan] False positive for struct member array element access

### ISSUE THREAD

**Issue by sfc-gh-oisler** [llvm/llvm-project#210643]
# `struct A { int elems[3]; }`

#### Code:
```cpp
struct A {
    int elems[3];
};

A a;

int main() {
    a.elems[0] = 1;
}
```

#### Logs:
```text
ASM generation compiler returned: 0
Execution build compiler returned: 0
Program returned: 0
==1==ERROR: TypeSanitizer: type-aliasing-violation on address 0x5ab0a1080ee4 (pc 0x5ab0a0726f87 bp 0x7ffeaab9afa0 sp 0x7ffeaab9af30 tid 1)
WRITE of size 4 at 0x5ab0a1080ee4 with type int accesses an existing object of type A
    #0 0x5ab0a0726f86  (/app/output.s+0x35f86)
```

#### Explanation:
A global object of type `A` contains an `int[3]` member; `a.elems[0]` designates an `int` subobject of `a`, and writing to that subobject (`a.elems[0] = 1;`) is valid C++ and does not violate aliasing rules. TypeSanitizer incorrectly reports this write as a type-aliasing violation on the enclosing `A` object.
